testbed/testbed/data_providers.go

```diff
@@ -308,3 +316,97 @@ func populateSpansMap(resourceSpansList []*otlptrace.ResourceSpans) map[string]*
+
+// FileDataProvider in an implementation of the DataProvider for use in performance tests.
+// The data to send is loaded from a file. The file should contain one JSON-encoded
+// Export*ServiceRequest Protobuf message. The file can be recorded using the "file"
+// exporter (note: "file" exporter writes one JSON message per line, FileDataProvider
+// expects just a single JSON message in the entire file).
+type FileDataProvider struct {
+	batchesGenerated   *atomic.Uint64
+	dataItemsGenerated *atomic.Uint64
+	message            proto.Message
+	ItemsPerBatch      int
+}
+
+// NewFileDataProvider creates an instance of FileDataProvider which generates test data
+// loaded from a file.
+func NewFileDataProvider(filePath string, dataType configmodels.DataType) (*FileDataProvider, error) {
+	file, err := os.OpenFile(filePath, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	var message proto.Message
+	var dataPointCount int
+
+	// Load the message from the file and count the data points.
+
+	switch dataType {
+	case configmodels.TracesDataType:
+		var msg otlptracecol.ExportTraceServiceRequest
+		if err := protobufJSONUnmarshaler.Unmarshal(file, &msg); err != nil {
+			return nil, err
+		}
+		message = &msg
+
+		md := pdata.TracesFromOtlp(msg.ResourceSpans)
+		dataPointCount = md.SpanCount()
+
+	case configmodels.MetricsDataType:
+		var msg otlpmetricscol.ExportMetricsServiceRequest
+		if err := protobufJSONUnmarshaler.Unmarshal(file, &msg); err != nil {
+			return nil, err
+		}
+		message = &msg
+
+		md := pdata.MetricsFromOtlp(msg.ResourceMetrics)
+		_, dataPointCount = md.MetricAndDataPointCount()
+
+	case configmodels.LogsDataType:
+		var msg otlplogscol.ExportLogsServiceRequest
+		if err := protobufJSONUnmarshaler.Unmarshal(file, &msg); err != nil {
+			return nil, err
+		}
+		message = &msg
+
+		md := pdata.LogsFromInternalRep(internal.LogsFromOtlp(msg.ResourceLogs))
+		dataPointCount = md.LogRecordCount()
+	}
+
+	return &FileDataProvider{
+		message:       message,
+		ItemsPerBatch: dataPointCount,
+	}, nil
+}
+
+func (dp *FileDataProvider) SetLoadGeneratorCounters(batchesGenerated *atomic.Uint64, dataItemsGenerated *atomic.Uint64) {
+	dp.batchesGenerated = batchesGenerated
+	dp.dataItemsGenerated = dataItemsGenerated
+}
+
+// Marshaler configuration used for marhsaling Protobuf to JSON. Use default config.
+var protobufJSONUnmarshaler = &jsonpb.Unmarshaler{}
+
+func (dp *FileDataProvider) GenerateTraces() (pdata.Traces, bool) {
+	// TODO: implement similar to GenerateMetrics.
+	return pdata.NewTraces(), true
+}
+
+func (dp *FileDataProvider) GenerateMetrics() (pdata.Metrics, bool) {
+	md := pdata.MetricsFromOtlp(dp.message.(*otlpmetricscol.ExportMetricsServiceRequest).ResourceMetrics)
+	dp.batchesGenerated.Inc()
+	_, dataPointCount := md.MetricAndDataPointCount()
+	dp.dataItemsGenerated.Add(uint64(dataPointCount))
+	return md, false
+}
+
+func (dp *FileDataProvider) GetGeneratedSpan(pdata.TraceID, pdata.SpanID) *otlptrace.Span {
```

Comment:
Interesting API :)) Can you file an issue to investigate this?

Reply:
Fixed the comment to explain. It is only used for data providers for correctness tests.